rust/yarp/build.rs

```diff
@@ -0,0 +1,507 @@
+use serde::Deserialize;
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+
+#[derive(Debug, Deserialize)]
+enum NodeParamType {
+    #[serde(rename = "node")]
+    Node,
+
+    #[serde(rename = "node?")]
+    OptionalNode,
+
+    #[serde(rename = "node[]")]
+    NodeList,
+
+    #[serde(rename = "string")]
+    String,
+
+    #[serde(rename = "constant")]
+    Constant,
+
+    #[serde(rename = "constant[]")]
+    ConstantList,
+
+    #[serde(rename = "location")]
+    Location,
+
+    #[serde(rename = "location?")]
+    OptionalLocation,
+
+    #[serde(rename = "location[]")]
+    LocationList,
+
+    #[serde(rename = "uint32")]
+    UInt32,
+
+    #[serde(rename = "flags")]
+    Flags
+}
+
+#[derive(Debug, Deserialize)]
+struct NodeParam {
+    name: String,
+    r#type: NodeParamType,
+    kind: Option<String>
+}
+
+#[derive(Debug, Deserialize)]
+struct Node {
+    name: String,
+    child_nodes: Option<Vec<NodeParam>>,
```

Comment:
I'm not sure if it's typically considered an anti-pattern or not, but, while `Option<Vec<T>>` does convey info to the developer (sometimes there'll be nodes and sometimes not), in practice it ends up being the same amount code to work with a `Vec<T>`. I could be wrong, but I also don't believe the compiler does any optimizing if you use `Option<Vec<T>>` over `Vec<T>`.

If your concern is with deserialization (i.e. sometimes there's a `child_nodes` field, sometimes not), you could use the `#[serde(default)]` attribute... https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6ed6ea58b8d6fc86f1e634d6b152e0a4

(sorry, if you're already familiar with all this and it's all intentional; I'll stop if that's the case!)